py/dml/compat.py

```diff
@@ -0,0 +1,186 @@
+features: dict[str, dict[str, CompatFeature]] = {
+    api: {} for api in env.api_versions().values()}
+
+
+def feature(cls: type[CompatFeature]):
+    assert issubclass(cls, CompatFeature)
+    singleton = cls()
+    features[cls.last_api_version][singleton.tag()] = singleton
+    return singleton
+
+
+@feature
+class port_proxy_ifaces(CompatFeature):
+    '''Version 5 and earlier of Simics relied on interface ports (as
+    registered by the `SIM_register_port_interface` API function) for
+    exposing the interfaces of ports and banks. In newer versions of
+    Simics, interfaces are instead exposed on separate configuration
+    objects.  When this feature is enabled, old-style
+    interface ports are created as proxies to the interfaces on the
+    respective port objects. Such proxies are not created for all
+    banks and ports; e.g., banks inside groups were not allowed in
+    Simics 5, so for such banks there is no old script that needs
+    proxies for compatibility.
+    '''
+    short = "Don't generate proxy port interfaces for banks and ports"
+    last_api_version = 6
+
+
+@feature
+class port_proxy_attrs(CompatFeature):
+    '''In Simics 5, configuration attributes for `connect`,
+    `attribute` and `register` objects inside banks and ports were
+    register on the device object, named like
+    <code><em>bankname</em>_<em>attrname</em></code>. Such proxy
+    attributes are only created When this feature is enabled.
+    Proxy attributes are not created for all banks and ports, in the
+    same manner as documented in the `port_proxy_ifaces` feature.
+    '''
+    short = ("Don't generate top-level proxy attributes"
+             + " for attributes in banks and ports")
+    last_api_version = 6
+
+
+@feature
+class io_memory(CompatFeature):
+    '''The `transaction` interface was introduced in 6, and will
+    eventually replace the `io_memory` interface. When this feature is
+    enabled, the top-level parameter `use_io_memory` defaults to
+    `true`, causing `bank` objects to implement `io_memory` instead of
+    `transaction` by default.'''
+    short = 'Use the io_memory interface by default in banks'
+    last_api_version = 6
+
+
+@feature
+class port_obj_param(CompatFeature):
+    '''This compatibility feature changes the value of the `obj`
+    parameter in `bank` and `port` objects: Before Simics 6 there were
+    no dedicated port objects, so this parameter did not exist and if
+    you wrote `obj` inside a bank, this would resolve to
+    `dev.obj`. This feature preserves this legacy behaviour by making
+    the `obj` parameter of banks and ports resolves to `dev.obj`
+    rather than the port object.
+    '''
+    short = "bank.obj and port.obj resolve to dev.obj"
+    last_api_version = 5
+
+
+@feature
+class dml12_inline(CompatFeature):
+    '''When inlining a method using `inline` in a DML 1.2 device,
+    constant parameters passed in typed arguments are also inlined,
+    unless compiling with `-g`.
```

Comment:
A dedicated tag is in case it has negative performance implications. And, admittedly, because it was one of the first tags I added, I later realized that most things should fit within misc.